feat: add queue action (#1729)
Mergifyio/mergify-engine: docs/source/queues.rst

etc...

Now, pull request ``D`` is embarked in merge train, ``B`` have been merged in
the base branch, its head commit have moved to ``12345678``. Mergify will creates:

* ``Z`` with the content of ``12345678`` + ``C`` + ``D``

If anything is not handled by Mergify is merged in the base branch, the merge
train is reset and the process start over.

Working with Queue Rules
~~~~~~~~~~~~~~~~~~~~~~~~

With a ``queue`` action, a pull request has to match two sets of conditions
to get merged. The first set of conditions described in ``pull_request_rules``
allows the pull request to enter in the queue when all conditions match. If
the ``queue`` action conditions do not match anymore the pull request is
removed from the merge queue. When a pull request is the next one to be
merged, Mergify looks at the seconds set of conditions defined this time in
``queue_rules`` and wait that they all match to merge the pull request.

When multiple queues are defined, they are processed one after the other in the
order they are defined in the configuration. When the first pull request in the
first queue is waiting to be merged, the remaining queues are blocked.

> 

Reviewer: we should add something about the cases where the PR is removed from the queue: either by being merged or by not matching the conditions anymore.

Mergify therefore creates the following pull requests:

* ``W`` with the content of ``abcdef12`` + ``A``
* ``X`` with the content of ``abcdef12`` + ``A`` + ``B``
* ``Y`` with the content of ``abcdef12`` + ``A`` + ``B`` + ``C``

The upside of creating multiple pull requests is that continuous integration
pipelines can run on all of them in parallel. Currently, Mergify embarks up to
5 pull requests into the merge train.

When the pull request ``W`` matches the ``queue_rules``, the pull request ``A``
is merged and the pull request ``W`` is deleted. The same goes for pull request
``X`` and ``B``, as well as for the pull request ``Y`` and ``C``.

At this point, let's say that ``A`` and ``B`` have been merged successfully.

If a 4th pull request ``D`` is created and matches the condition for the queue
action, it is also embarked into the merge train. Since ``B`` have been merged
in the base branch, the new head commit for the base branch is ``12345678``.

In that case, Mergify creates a new pull request ``Z`` with the content of
``12345678`` + ``C`` + ``D``.

In the case where a new commit is pushed on the base branch, Mergify resets the
merge train and starts over the process.

> If a pull request embarked doesn't match the ``queue_rules`` anymore, this one
> is retired from the merge train and all pull requests embarked after this
> one are dis-embarked and re-embarked.

Reviewer: ```suggestion
If an embarked pull request doesn't match the ``queue_rules`` anymore, it
is removed from the merge train. All pull requests embarked after it
are disembarked and re-embarked.
```

Reviewer: Can you just update this ^ ?

Working with Queue Rules
~~~~~~~~~~~~~~~~~~~~~~~~

With a ``queue`` action, a pull request has to match two sets of conditions
to get merged. The first set of conditions described in ``pull_request_rules``
allows the pull request to enter in the queue when all conditions match. If
the ``queue`` action conditions do not match anymore the pull request is

> removed from the merge queue. When a pull request is the next one to be
> merged, Mergify looks at the seconds set of conditions defined this time in

Reviewer: ```suggestion
removed from the merge queue. When a pull request is at the top of the queue, Mergify looks at the seconds set of conditions defined this time in
```
